### PR TITLE
Feature find intersected elements

### PR DESCRIPTION
**Pull request by pooyan-dadvand**
This PR adds a new `FindIntersectedGeometricalObjectsProcess` class and include tests for it. There are two notes to be commented:
- I had to add the Intersection method to triangle 3d3 which was missing. This implementation is borrowed from a reference and I did some refactoring but still needs more. However it was out of scope of this branch.

- There are some lines commented in GiDIO which I need them for my tests in another branches. I will remove them when I finish the work on distances

### CODE REVIEW

File: kratos/processes/find_intersected_geometrical_objects_process.cpp [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+		// loop over all skin nodes
+		for (auto i_node = mrModelPart2.NodesBegin(); i_node != mrModelPart2.NodesEnd(); i_node++) {
+			for (int i = 0; i < 3; i++) {
+				low[i] = i_node->Coordinate(i + 1) < low[i] ? i_node->Coordinate(i + 1) : low[i];
+				high[i] = i_node->Coordinate(i + 1) > high[i] ? i_node->Coordinate(i + 1) : high[i];
+			}
+		}
+
+		// TODO: Octree needs refactoring to work with BoundingBox. Pooyan.
```
Comment (RiccardoRossi) [inline kratos/processes/find_intersected_geometrical_objects_process.cpp]: .data.data ?

just one comment, don't you want to subtract an epsilon from the minimum and add it to the max?

File: kratos/tests/processes/test_calculate_distance_to_skin_process.cpp [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+      Parameters mesher_parameters(R"(
+            {
+                "number_of_divisions":   10,
+                "element_name":     "Element3D4N"
+            })");
+
+      ModelPart volume_part("Volume");
+      volume_part.AddNodalSolutionStepVariable(VELOCITY);
+      volume_part.AddNodalSolutionStepVariable(DISTANCE);
+      volume_part.AddNodalSolutionStepVariable(EMBEDDED_VELOCITY);
```
Comment (RiccardoRossi) [inline kratos/tests/processes/test_calculate_distance_to_skin_process.cpp:50]: i would say EMBEDDED_VELOCITY is an elemental variable, not a nodal one

**Review by RiccardoRossi — COMMENTED** [review]
i can not see anything wrong, just minor comments, however i did not compile the pull request since i am in windows

File: kratos/tests/processes/test_calculate_distance_to_skin_process.cpp (hunk 2 above)
Comment (pooyan-dadvand) [reply kratos/tests/processes/test_calculate_distance_to_skin_process.cpp:50]: @rubenzorrilla : is it true?

Comment (rubenzorrilla) [reply kratos/tests/processes/test_calculate_distance_to_skin_process.cpp:50]: Yes, he is right. My bad...

File: kratos/processes/find_intersected_geometrical_objects_process.cpp (hunk 1 above)
Comment (pooyan-dadvand) [reply kratos/processes/find_intersected_geometrical_objects_process.cpp]: The first `data()` is from `array_1d` and returns the `boost::array` which is the object of second one. I  agree that it is NOT clean at all. Nevertheless I would postpone this to when we refactor the spatial searches.

Comment (pooyan-dadvand) [reply kratos/processes/find_intersected_geometrical_objects_process.cpp]: @RiccardoRossi about the epsilon in which line you mean?

File: kratos/processes/find_intersected_geometrical_objects_process.cpp [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+		}
+
+		// loop over all skin nodes
+		for (auto i_node = mrModelPart2.NodesBegin(); i_node != mrModelPart2.NodesEnd(); i_node++) {
+			for (int i = 0; i < 3; i++) {
+				low[i] = i_node->Coordinate(i + 1) < low[i] ? i_node->Coordinate(i + 1) : low[i];
+				high[i] = i_node->Coordinate(i + 1) > high[i] ? i_node->Coordinate(i + 1) : high[i];
+			}
+		}
+
```
Comment (RiccardoRossi) [inline kratos/processes/find_intersected_geometrical_objects_process.cpp:94]: here you are computing the maximal and minimal coordinates to identify the bounding box.

before assigning them, (so around this line) i would do
low[0] -= eps
low[1] -= eps
..

high[0] += eps

...

this is just to ensure that there are no problems with nodes or elements EXACTLY on the boundary

cheers
Riccardo

**Review by RiccardoRossi — APPROVED** [review]
@pooyan-dadvand to me the thing is "approved"...

File: kratos/processes/find_intersected_geometrical_objects_process.cpp (hunk 3 above)
Comment (pooyan-dadvand) [reply kratos/processes/find_intersected_geometrical_objects_process.cpp:94]: Thanks! Your are right! I have just forgotten to add it. It is added now.

**Pull request merged**